### PR TITLE
Build indicesAccessControl for selective indices

### PR DESCRIPTION
**Pull request by ywangd**
During index authorization, the IndicesAccessControl is built for all
original indices. The list of original indices can be large for a large
size cluster and requests with liberal wildcards. Therefore computing
IndicesAccessControl for every one of them can be expensive.

The observation is that many shard/index level request only really just
target a single concrete index and hence just need IndicesAccessControl
for that single index. One twist is that the single index may be
requested with its alias or authorized as part of a data stream. In this
case, the relevant aliases and data stream also need to be considered.
Still this list of concrete index and its aliases and data stream can be
much smaller than the original indices.

This PR adds an effective allowlist for requests that we know how to
extract the single concrete target index and only compute
IndicesAccessControl for the index and relevant aliases, data stream.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 1]
```diff
@@ -626,14 +630,45 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
...
+    private String getSingleTargetIndex(TransportRequest request) {
+        if (request instanceof ShardSearchRequest) {
+            return ((ShardSearchRequest) request).shardId().getIndexName();
+        } else if (request instanceof FieldCapabilitiesIndexRequest) {
+            return ((FieldCapabilitiesIndexRequest) request).index();
+        } else if (request instanceof ShardFetchSearchRequest) {
+            return ((ShardFetchSearchRequest) request).getShardSearchRequest().shardId().getIndexName();
+        }
+        return null;
+    }
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:670]: There are more Transport Requests that target a single index/shard. These three are chosen to match what are currently being used in benchmarking. They also serve the purpose of demostrating the general idea of this change. I'll work on a more complete list once we agree on the approach.

**Review by albertzaharovits — COMMENTED** [review]
I took a very close look at this.
My hope is that the same improvement is going to be covered in core, by https://github.com/elastic/elasticsearch/pull/78508 . I think that is almost there, and it also avoids the serialization/deserialization costs you were mentioning in the last weekly meeting.

**Comment by ywangd** [comment]
> I took a very close look at this. My hope is that the same improvement is going to be covered in core, by #78508 . I think that is almost there, and it also avoids the serialization/deserialization costs you were mentioning in the last weekly meeting.

Thanks Albert. Filtering the list of names in the core (#78508) is a superior solution in terms of performance. It covers this PR and more:
1. It does what this PR is trying to do in a more natural way
2. It reduces the looping cost (due to small set of requested indices) in `IndicesAndAliasesResolver#resolveIndicesAndAliases`
3. More importantly, it cut downs the de/serialisation cost for sending requests cross nodes. 

I am happy to close this PR once #78508 is merged. We should also discuss whether the same change should be actively promoted to other parts in core where applicable. I think it should be. But discussion would be great to understand the impact better and also help decide on whether core should always be aware of the source names, i.e. whether the concrete index name is from an alias or even multiple aliases. 

PS: This PR and other hacks locally almost screamed for the change in #78508. But I just didn't notice it. A lesson learned on how to tackle an issue in a more broader context (than just security).

**Comment by ywangd** [comment]
Closing since changes in the core (#78508) is a better pattern.

**Pull request closed without merging**